camel/tasks/task.py

```diff
@@ -0,0 +1,281 @@
+                if subtask.state != TaskSate.DELETED:
+                    subtask.set_state(state)
+        elif state == TaskSate.RUNNING:
+            if self.parent is not None:
+                self.parent.set_state(state)
+
+    def add_subtask(self, task: "Task"):
+        task.parent = self
+        self.subtasks.append(task)
+
+    def get_running_task(self) -> Optional["Task"]:
+        for sub in self.subtasks:
+            if sub.state == TaskSate.RUNNING:
+                return sub.get_running_task()
+        if self.state == TaskSate.RUNNING:
+            return self
+        return None
+
+    def to_string(self, indent="") -> str:
+        _str = f"{indent}Task {self.id}: {self.content}\n"
+        for subtask in self.subtasks:
+            _str += subtask.to_string(indent + "  ")
+        return _str
+
+
+def parse_response(response: str, task_id: Optional[str] = None) -> List[Task]:
+    pattern = "<task>(.*?)</task>"
+    tasks_content = re.findall(pattern, response, re.DOTALL)
+
+    tasks = []
+    if task_id is None:
+        task_id = "0"
+    for i, content in enumerate(tasks_content):
+        tasks.append(Task(content=content.strip(), id=f"{task_id}.{i}"))
+    return tasks
+
+
+class TaskManager:
+    """
+    TaskManager is used to manage tasks.
+
+    Attributes:
+        root_task: The root task.
+        tasks: The ordered tasks.
+        task_map: A map for task.id to Task.
+        current_task_id: The current "RUNNING" task.id.
+
+    Args:
+        task (Task): The root Task.
+    """
+
+    def __init__(self, task: Task):
+        self.root_task: Task = task
+        self.current_task_id: str = task.id
+        self.tasks: List[Task] = [task]
+        self.task_map: Dict[str, Task] = {task.id: task}
+
+    def exist(self, task_id: str) -> bool:
+        return task_id in self.task_map
+
+    @property
+    def current_task(self) -> Optional[Task]:
+        return self.task_map.get(self.current_task_id, None)
+
+    @staticmethod
+    def topological_sort(tasks: List[Task]) -> List[Task]:
+        stack = []
+        visited = set()
+
+        # recursive visit the vertices
+        def visit(task: Task):
+            if task.id in visited:
+                return
+            visited.add(task.id)
+
+            # go deep for dependencies
+            for sub_task in task.subtasks:
+                visit(sub_task)
+
+            # add current task to stack which have no dependencies.
+            stack.append(task)
+
+        for task in tasks:
+            visit(task)
+
+        return stack
+
+    def add_tasks(self, tasks: Union[Task, List[Task]]) -> None:
+        r"""
+        self.tasks and self.task_map will be updated by the input tasks.
+        """
+        if not tasks:
+            return
+        if not isinstance(tasks, List):
+            tasks = [tasks]
+        for task in tasks:
+            assert not self.exist(task.id), f"`{task.id}` already existed."
```

Comment:
I was thinking that for the `Workforce` collaboration module, maybe we don't need to manage the task `id`s ourselves, but when there are new tasks, we specify the parent node of the new tasks, and then the task automatically assigns ids to them based on the number of tasks we have, and returns the `id`. What do you think?

Reply:
The task id is assigned when it's created. if you do not want to manage the taskid, May it can be set as optional. But you need to set the parent of sub-tasks.

Reply:
ID setting as optional is OK. As we discuss in the dev meeting, I also hope `workforce` module can know the order of subtasks without no need to manage their IDs also.